### PR TITLE
crimson/osd: bring EC writes and reads with crimson as primary

### PR DESCRIPTION
**Pull request by rzarzynski**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/osd/ECBackend.cc [hunk 1]
```diff
@@ -275,19 +275,20 @@ struct OnRecoveryReadComplete :
 
 struct RecoveryMessages {
   map<hobject_t,
-      ECBackend::read_request_t> reads;
+      ECBackend::read_request_t> recovery_reads;
   map<hobject_t, set<int>> want_to_read;
-  void read(
+  void recovery_read(
```
Comment (tchaikov) [inline src/osd/ECBackend.cc:162]: i'd suggest s/grep/[lsp](https://langserver.org/)/, if you are not using it yet. personally, i've been using both clangd and ccls along with my editor's integrations. they work great under most circumstances.

**Review by athanatos — APPROVED** [review]
This also looks reasonable to me.  I'm inclined to get this merged sooner rather than later as conflicts seem likely.  We can handle cleanups in subsequent PRs.

**Comment by rzarzynski** [comment]
Rebasing.

**Comment by rzarzynski** [comment]
Rebased on top of the shared EC infrastructure.

**Comment by rzarzynski** [comment]
Fixed the Clang-noticed unused lambda capture.

**Comment by ljflores** [comment]
Passed QA, but looks like this needs to be rebased:
https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues65126

**Comment by yuriw** [comment]
@rzarzynski This is ready for merge after rebase
re: https://tracker.ceph.com/issues/65126

**Pull request closed without merging**